frame/vamm/src/helpers/checks.rs

```diff
@@ -0,0 +1,128 @@
+use crate::{Config, Error, Pallet, VammStateOf};
+use composable_traits::vamm::{AssetType, Direction, SwapConfig, SwapOutput};
+use frame_support::pallet_prelude::*;
+use sp_runtime::traits::{CheckedAdd, Zero};
+
+impl<T: Config> Pallet<T> {
+	/// Checks if the following properties hold before performing a swap:
+	///
+	/// * Vamm is open.
+	/// * There is a sufficient amount of assets in the reserves to give to the
+	/// caller if the swap is a [`Remove`](Direction::Remove) operation.
+	/// * The total amount of assets in the reserve will not overflow if the
+	/// swap is a [`Add`](Direction::Add) operation.
+	///
+	/// # Errors
+	///
+	/// * [`Error::<T>::VammIsClosed`]
+	/// * [`Error::<T>::InsufficientFundsForTrade`]
+	/// * [`Error::<T>::TradeExtrapolatesMaximumSupportedAmount`]
+	pub fn sanity_check_before_swap(
+		// config: &SwapConfigOf<T>,
+		config: &SwapConfig<T::VammId, T::Balance>,
+		vamm_state: &VammStateOf<T>,
+	) -> Result<(), DispatchError> {
+		// We must ensure that the vamm is not closed before performing any swap.
+		ensure!(!Self::is_vamm_closed(vamm_state, &None), Error::<T>::VammIsClosed);
+
+		match config.direction {
+			// If we intend to remove some asset amount from vamm, we must
+			// have sufficient funds for it.
+			Direction::Remove => match config.asset {
+				AssetType::Base => ensure!(
+					config.input_amount < vamm_state.base_asset_reserves,
+					Error::<T>::InsufficientFundsForTrade
+				),
+				AssetType::Quote => ensure!(
+					config.input_amount < vamm_state.quote_asset_reserves,
+					Error::<T>::InsufficientFundsForTrade
+				),
+			},
+
+			// If we intend to add some asset amount to the vamm, the
+			// final amount must not overflow.
+			Direction::Add => match config.asset {
+				AssetType::Base => ensure!(
+					config.input_amount.checked_add(&vamm_state.base_asset_reserves).is_some(),
+					Error::<T>::TradeExtrapolatesMaximumSupportedAmount
+				),
+				AssetType::Quote => ensure!(
+					config.input_amount.checked_add(&vamm_state.quote_asset_reserves).is_some(),
+					Error::<T>::TradeExtrapolatesMaximumSupportedAmount
+				),
+			},
+		};
+
+		Ok(())
+	}
+
+	/// Checks if the following properties hold after performing a swap:
+	///
+	/// * Swapped amount respects limit specified in
+	/// [`SwapConfig::output_amount_limit`].
+	/// * Base asses was not completely drained.
```

Comment:
Base assets* :)

Reply:
"Base asses was not completely drained" is a great quote

Reply:
LOL

:see_no_evil: :see_no_evil: :see_no_evil: 

FIXED :+1: